hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/DeltaSync.java

```diff
@@ -451,6 +457,18 @@ public void refreshTimeline() throws IOException {
     return Pair.of(schemaProvider, Pair.of(checkpointStr, records));
   }
 
+  protected Option<String> getPreviousCheckpoint(HoodieTimeline timeline) throws IOException {
+    return timeline.getReverseOrderedInstants().map(instant -> {
+      try {
+        HoodieCommitMetadata commitMetadata = HoodieCommitMetadata
+            .fromBytes(timeline.getInstantDetails(instant).get(), HoodieCommitMetadata.class);
+        return Option.ofNullable(commitMetadata.getMetadata(CHECKPOINT_KEY));
+      } catch (IOException e) {
+        throw new HoodieIOException("Failed to parse HoodieCommitMetadata for " + instant.toString(), e);
+      }
+    }).filter(Option::isPresent).findFirst().orElse(Option.empty());
```

Comment:
Is this lazily evaluated? Just verifying because if not then won't this always walk back every commit and then discard all except the latest? That will be a lot for large existing tables.